### ISSUE TITLE
Bedrock users can't create Town when logging in via Geyser MC/floodgate due to not being 'registered'

### ISSUE THREAD

**Issue by chairmanmow** [TownyAdvanced/Towny#4664]
### What steps will reproduce the problem?
1. Setup a working SpigotMC Server using GeyserMC/Floodgate so that Bedrock users can connect
2. Connect as a bedrock user - Server throws exception in Console from Towny

`[01:22:16] [Network Listener - #1/INFO]: [Geyser-Spigot] /24.247.183.100:55466 tried to connect!
[01:22:17] [Network Listener - #1/INFO]: [Geyser-Spigot] Player connected with username DoctorDer5402
[01:22:23] [Network Listener - #1/INFO]: [Geyser-Spigot] Attempting to login using floodgate mode... authentication will be encrypted.
[01:22:23] [Thread-64/WARN]: [Geyser-Spigot] Loaded class org.geysermc.floodgate.util.EncryptionUtil from floodgate-bukkit v1.0-SNAPSHOT which is not a
 depend, softdepend or loadbefore of this plugin.
[01:22:23] [Thread-64/INFO]: [Geyser-Spigot] Loaded Floodgate key!
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] DoctorDer5402 (logged in as: DoctorDer5402) has connected to remote java server on address 127
.0.0.1
[01:22:23] [nioEventLoopGroup-2-2/WARN]: [floodgate-bukkit] Loaded class org.geysermc.floodgate.util.BedrockData from Geyser-Spigot v1.2.0-SNAPSHOT whi
ch is not a depend, softdepend or loadbefore of this plugin.
[01:22:23] [Netty Epoll Server IO #2/INFO]: [floodgate-bukkit] Added *DoctorDer5402 00000000-0000-0000-0009-01fce649dc69
[01:22:23] [Netty Epoll Server IO #2/INFO]: UUID of player *DoctorDer5402 is 00000000-0000-0000-0009-01fce649dc69
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] Registering bedrock skin for *DoctorDer5402 (00000000-0000-0000-0009-01fce649dc69)
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] Unable to load bedrock skin for '*DoctorDer5402' as they are likely using a customised skin
[01:22:23] [Server thread/INFO]: *DoctorDer5402 joined the game
[01:22:24] [Server thread/INFO]: *DoctorDer5402[/24.247.183.100:35640] logged in with entity id 3901 at ([world]-2448.5, 64.0, -1249.5)
[01:22:24] [Server thread/WARN]: [Towny] Task #12199 for Towny v0.96.5.0 generated an exception
java.lang.NullPointerException: null
        at com.palmergames.bukkit.towny.tasks.OnPlayerLogin.run(OnPlayerLogin.java:121) ~[?:?]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftTask.run(CraftTask.java:99) ~[patched_1.16.4.jar:git-Paper-376]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:468) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1294) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:377) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1209) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:997) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:178) ~[patched_1.16.4.jar:git-Paper-376]
        at java.lang.Thread.run(Thread.java:834) [?:?]
[01:22:24] [Thread-65/INFO]: [Geyser-Spigot] Spawned player at -2,448.5 64 -1,249.5
[01:22:27] [Thread-65/INFO]: [Geyser-Spigot] Spawned player at -2,448.5 64 -1,249.5`

3.  From the user perspective, once they try to enter `/town create sometown` they would see this error on their end `[Towny] The resident '*DoctorDer5402' is not registered`

### What is the expected output?
1. Logging in as bedrock users works, however, user's in the same groups (same permissions) can't create towns while Java users can.  
2. No NullPointerException is thrown in the console when a user logs in from a Bedrock account via GeyserMC/Floodgate from the Towny `onPlayerLogin` event handler and registration succeeds
3. From the user perspective, they are able to proceed onward through town creation like a Java user from the same permission group.  

### What is your Towny version?
Towny 0.96.5

### What is your spigot/craftbukkit version?
paper-1.16.4-376

### Please use Pastebin.com to link the following files:

**Skipping these files is not an option. It is the easiest way to diagnose an issue ticket properly. If you do skip this step then your ticket will most likely be deleted and you will be asked to resubmit.**
1. Your full server startup from the logs\latest.log [https://pastebin.com/S3EBfY5a](https://pastebin.com/S3EBfY5a)

2. Your Towny config.yml (if using MySQL please remove password,) [https://pastebin.com/ve9vBbXx](https://pastebin.com/ve9vBbXx)

3. Your townyperms.yml : [https://pastebin.com/vQyES895](https://pastebin.com/vQyES895)

4. Your log's error : 

`[01:22:16] [Network Listener - #1/INFO]: [Geyser-Spigot] /24.247.183.100:55466 tried to connect!
[01:22:17] [Network Listener - #1/INFO]: [Geyser-Spigot] Player connected with username DoctorDer5402
[01:22:23] [Network Listener - #1/INFO]: [Geyser-Spigot] Attempting to login using floodgate mode... authentication will be encrypted.
[01:22:23] [Thread-64/WARN]: [Geyser-Spigot] Loaded class org.geysermc.floodgate.util.EncryptionUtil from floodgate-bukkit v1.0-SNAPSHOT which is not a
 depend, softdepend or loadbefore of this plugin.
[01:22:23] [Thread-64/INFO]: [Geyser-Spigot] Loaded Floodgate key!
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] DoctorDer5402 (logged in as: DoctorDer5402) has connected to remote java server on address 127
.0.0.1
[01:22:23] [nioEventLoopGroup-2-2/WARN]: [floodgate-bukkit] Loaded class org.geysermc.floodgate.util.BedrockData from Geyser-Spigot v1.2.0-SNAPSHOT whi
ch is not a depend, softdepend or loadbefore of this plugin.
[01:22:23] [Netty Epoll Server IO #2/INFO]: [floodgate-bukkit] Added *DoctorDer5402 00000000-0000-0000-0009-01fce649dc69
[01:22:23] [Netty Epoll Server IO #2/INFO]: UUID of player *DoctorDer5402 is 00000000-0000-0000-0009-01fce649dc69
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] Registering bedrock skin for *DoctorDer5402 (00000000-0000-0000-0009-01fce649dc69)
[01:22:23] [nioEventLoopGroup-2-2/INFO]: [Geyser-Spigot] Unable to load bedrock skin for '*DoctorDer5402' as they are likely using a customised skin
[01:22:23] [Server thread/INFO]: *DoctorDer5402 joined the game
[01:22:24] [Server thread/INFO]: *DoctorDer5402[/24.247.183.100:35640] logged in with entity id 3901 at ([world]-2448.5, 64.0, -1249.5)
[01:22:24] [Server thread/WARN]: [Towny] Task #12199 for Towny v0.96.5.0 generated an exception
java.lang.NullPointerException: null
        at com.palmergames.bukkit.towny.tasks.OnPlayerLogin.run(OnPlayerLogin.java:121) ~[?:?]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftTask.run(CraftTask.java:99) ~[patched_1.16.4.jar:git-Paper-376]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:468) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1294) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:377) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1209) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:997) ~[patched_1.16.4.jar:git-Paper-376]
        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:178) ~[patched_1.16.4.jar:git-Paper-376]
        at java.lang.Thread.run(Thread.java:834) [?:?]
[01:22:24] [Thread-65/INFO]: [Geyser-Spigot] Spawned player at -2,448.5 64 -1,249.5
[01:22:27] [Thread-65/INFO]: [Geyser-Spigot] Spawned player at -2,448.5 64 -1,249.5
`

**Comment by LlmDl** [comment]
I am not paid enough to worry about geyser servers that use floodgate, sorry.

**Comment by chairmanmow** [comment]
@LlmDl Sorry, don't mean to cause you worry, nor suggest you should take up any unpaid work on account of this issue.  Are you the only one working on this project?  Assuming this is open source not sure why you need to close it; someone else can take it up and submit a PR, or at least can be discussed a bit - as is you've closed the issue even though this bug report pretty much points to the exact line where you'd need to debug into.   Bug may be definitely mislabel, I apologize for that as it may be offensive, call it 'feature request' if that's more apt to be left open.  The bug is likely in Geyser-MC, but I'm guessing it may be easier to fix here.  

 I'm not a Java Dev nor have I examined the whole call stack but just based on the error looks like it could be one of two things causing it to me from the logs, either they are not conforming to the interface your method expects when it gets to the 'onPlayerLogon' or more likely to me the event is getting fired too soon with null data getting passed to whatever method handles 'registration'.  
 
In any event, I can understand the unwillingness to deal with either a non-conforming API or async issue that may arise outside your own code, however perhaps there is a way to allow Bedrock users to register without going through all those headaches of finding the root cause there - what I was thinking exposing some command like `/townyregister <params needed?>`  that would more or less wrap whatever method is being called  in the logonHandler to 'register' if a user is not already 'registered', the needed scope to handle generating the internal parameters is likely initialized by then.  Or perhaps another way to handle it is to simply handle at the point the user is shown '[Towny] The resident '*DoctorDer5402' is not registered` and try calling the 'registration' function again in that block.

If you're actually hurting for money and not joking about the potential pain of working with Geyser stuff and maybe if any of my suggestions above make sense, and you feel like working it we can do some sort of compensation arrangement if you're looking for incentive.   No worries if not though - but if I or someone else were able to figure out a way to tackle this do you prefer people maintain their own forks or would you appreciate PR's back to this repository?  If my machine and brain were setup and ready to go for Java Development think I could take a swing at this and possibly figure it out, maybe not - since I'm not though figured the repository a good first place to start.

Anyways, No rush or need to respond - is what it is for now and that's cool, it's a Minecraft plugin for fun after all.  I do want to get it running for Bedrock too but there are many more important things including your personal expense time/effort/money.  Maybe this follow-up offers more hope or maybe I'm missing something with what might be happening - hopefully I was able to offer some suggestions that may be reduce the scope or burden of the issue and if I can be of any other assistance on this issue or generally supporting the project feel free to let me know but otherwise, hopefully nothing to annoying in this follow-up comment; sincerest apologies if so! Cheers in any event!

**Comment by chairmanmow** [comment]
@LlmDl I decided to sponsor you cause I read your bio and saw money was tight so I guess you're not joking - I know it's not much but no expectations on you spending any time on this, but I think maybe I've got it figured out more or less on an architectural level, shouldn't hopefully take too long, I wouldn't want you to do any more than what I'm suggesting here as you are right it's probably not worth it for you.  

Anyways, I looked more at the code and it looks like the required parameters to essentially register are 'player name' and 'player UUID', and that the registration code is just this block : https://github.com/TownyAdvanced/Towny/blob/2e8f8fd0992b5fe0495f3bc46262460edd2a1739/src/com/palmergames/bukkit/towny/tasks/OnPlayerLogin.java#L75-L98

If for some reason you feel like wrapping that block or make a similar one inside a callable function that is callable via command, like '/townyregister' and either pick up the params playerName, UUID from command sender API or passed parameters via CLI, I think it would make things work.  If you have any questions/comments let me know

**Comment by LlmDl** [comment]
Hi, I haven't seen an email about the sponsorship yet but I will let you know. 

PRs are welcome as long as the code is good (accommodating a floodgate geyser server could get fairly janky,) and you can actually get a good UUID from it (I have seen other admins using geyser & floodgate get UUIDs for every bedrock player.)

**Comment by LlmDl** [comment]
Just to update you: I did get the email about the sponsorship, thanks.

**Comment by chairmanmow** [comment]
@LlmDl Glad to hear the donation worked looked like github had an issue with the CC info I put in, so I fixed it.  As far as Floodgate/Geyser giving me the same UUID per player, I am getting unique ID's for each one (if I wasn't there wouldn't be much that could be done, agreed), so that shouldn't be an issue for me at least.  There's a lookup tool somewhere to find the UUID from a Bedrock user as well, and those seemed to be matching 1:1 as I used those for whitelist functionality earlier.  I'd have to dig around for that lookup tool if you're curious, but it matched what I received on the server.  I mean it may not be the UUID that's tripping it up, it does add an '*' asterisk prefix to usernames during logon which I could imagine could causes wacky behavior but the explicit call would allow the user to register after the GeyserMC initializes (which I'm guessing happens after the 'onLogon' event fires)

So assuming I had a user with name '*SomeUserName' and a UUID of '00000000-0000-0000-0009-00000a941d10' would be able to run `/townyregister -user *SomeUserName -id 00000000-0000-0000-0009-00000a941d10` or optionally omit the parameters if those can get pulled from the message sender [user] scope.  

May take a look myself in the future and submit a PR, but would be some time as I don't have much space on my personal computer to set up a decent Java Dev Environment at the moment (although playing with Minecraft Servers is making this tempting to do).  However if you feel like taking a shot, would be appreciated.  Take care.

**Comment by LlmDl** [comment]
You might find this discussion post useful: https://github.com/TownyAdvanced/Towny/discussions/4568#discussioncomment-195458

**Comment by chairmanmow** [comment]
@LlmDl Looks like this has come up before, so you think it's the regEx for towny rejecting '*'?  That should be easy enough to fix with Geyser/floodgate as it's configurable in one of those places.  Does the regex allow underscores however (some Bedrock names can have spaces, represented as underscores when logging in)?    I could take a look at the regex, too, haven't tracked it down, at work at the moment, but later I will likely just reconfigure Geyser/floodgate to remove the '*' and not worry about name collision for now to see if that's the issue when I am done, let you know if that was the cause.

As an aside though relative to that comment, not sure I'd want to run Geyser without floodgate, as it's my understanding that once you're doing that, it would require Bedrock players to also have paid for a Java / Mojang license, at which point I wonder why'd they'd be running Bedrock - although I prefer playing games on console myself and using my computer for other things but I'm in the minority these days.

**Comment by chairmanmow** [comment]
Looks like underscores should be valid, I'll change the prefix from * to _ later and see

**Comment by LlmDl** [comment]
underscores are an allowed character in java account usernames while * is not.

**Comment by chairmanmow** [comment]
@LlmDl So it's working now, more or less - was able to run a quick test, it is the name prefix being an * that causes the issue to appear, setting `username-prefix: "_"` in `bukkit/config.yml` will allow things to function normally.  Created a new dummy Xbox user and was able to login, get no errors, and create a town.   

For me that's more or less fine, although I will need to clear out my user data so the '*' value doesn't appear with some UUIDS I've used (which seems to be caching from somewhere I can't find yet, separate issue I'll figure it out).  I suppose if you were running a server that already had a bunch of people up and running it would be an issue to repair but as long as I can clear it out I should be good once I find out where these cached values are coming from.  Should be all good now! 😃

**Comment by LlmDl** [comment]
good luck

**Comment by chairmanmow** [comment]
@LlmDl Thanks, you too!  I was actually able to repair existing users somewhat on a per user basis using what amounts to find and replace in Linux, maybe it helps someone else, if you were to change the bukkit/config.yml to change prefix from `*` to `_` for exampleUser '*BedrockGuy' :

`grep -Ril "*BedrockGuy" /opt/minecraft/spigotMC/ | xargs sed -i 's/\*BedrockGuy/\_BedrockGuy/g'`

Repeat for any user to be repaired.  My disclaimer : probably should make a backup if anyone reads this and doesn't understand exactly what they're doing...  Anyways, problem totally solved now.  Sorry it was an issue with the bukkit config in the first place, but probably some good info in this ticket at this point.

**Comment by Camotoy** [comment]
This was linked in one of Geyser's chats, so in case this is found by someone else I'll mention why the prefix didn't change for existing users.

Paper has a bug where users that connect with no texture have their properties overwritten by the usercache - so, if the name of such a player changes, the contents of `usercache.json` located at the server directory would be used instead. You simply need to delete that file and restart your server, and the issue would be gone. Once https://github.com/PaperMC/Paper/pull/4970 is merged or addressed, this will no longer be an issue.
